purchase_triple_discount/report/purchase_report.py

```diff
@@ -0,0 +1,42 @@
+# Copyright 2019 GRAP (http://www.grap.coop)
+# Sylvain LE GAL (https://twitter.com/legalsylvain)
+# License AGPL-3 - See http://www.gnu.org/licenses/agpl-3.0.html
+
+from odoo import fields, models
+import odoo.addons.decimal_precision as dp
+
+
+class PurchaseReport(models.Model):
+    _inherit = "purchase.report"
+
+    discount2 = fields.Float(
+        string='Discount 2 (%)', digits=dp.get_precision('Discount'),
+        group_operator="avg",
+    )
+
+    discount3 = fields.Float(
+        string='Discount 3 (%)', digits=dp.get_precision('Discount'),
+        group_operator="avg",
+    )
+
+    def _select(self):
+        res = super()._select()
+        res += ", l.discount2 AS discount2, l.discount3 AS discount3"
+        return res
+
+    def _group_by(self):
+        res = super()._group_by()
+        res += ", l.discount2, l.discount3"
+        return res
+
+    def _get_discounted_price_unit_exp(self):
+        """Inheritable method for getting the SQL expression used for
+        calculating the unit price with discount(s).
+
+        :rtype: str
+        :return: SQL expression for discounted unit price.
+        """
+        return """
+            (1.0 - COALESCE(l.discount, 0.0) / 100.0) *
```

Comment:
Can this be simplified somehow for better performance?

Reply:
What do you mean exactly ? 
In fact, I'm really newb in ``report`` features. (I don't use it, for GRAP, I only used materialized view with ``bi_sql_editor``). I wrote that part only to remove the roadmap part ;-)

is ``COALESCE`` requiring a lot of time to compute ? I don't see how to make it more simple.

Reply:
It was more the number of operations. I'm proposing a refactoring in short.